### PR TITLE
Improving detection of scroll ended

### PR DESCRIPTION
**Pull request by guilhermefloriani**
We are facing some issues like #697 #722 #322 #564 and few others, reporting that there is a problem related to the detection of scroll ended.

Like some people already said in this kind of issue, this is happening because the browser is delaying the `setTimeout` callback on `_debounceScrollEnded` in exchange of performance reasons.

This is a proposal to fix this kind of problems using `requestAnimationFrame` instead of `setTimeout` and it works fine for me.

This PR includes:
- Using `requestAnimationFrame` instead of `setTimeout` in `Grid` component and `Masonry` component.
- Fixing jest setup to polyfill `requestAnimationFrame` and `cancelAnimationFrame`.
- Change a `Masonry` test to wait before all "animations" finish before assert the result. (I did it with a `setTimeout` callback because `jest.runOnlyPendingTimers()` does not wait a task scheduled with `requestAnimationFrame`, and this kind of task have priority for `setTimeout` "tasks", so the assert expression will be executed only after `_debounceScrollEnded` was called).

### CODE REVIEW

File: source/Grid/Grid.js [hunk 1]
```diff
@@ -979,16 +979,20 @@ export default class Grid extends PureComponent {
...
-      clearTimeout(this._disablePointerEventsTimeoutId)
+      window.cancelAnimationFrame(this._disablePointerEventsTimeoutId)
     }
 
-    this._disablePointerEventsTimeoutId = setTimeout(
-      this._debounceScrollEndedCallback,
-      scrollingResetTimeInterval
-    )
+    const delay = () => {
+      if (Date.now() - this._scrollDebounceStart >= this.props.scrollingResetTimeInterval) {
```
Comment (bvaughn) [inline source/Grid/Grid.js]: We could assign `scrollingResetTimeInterval` to a local variable and avoid the `this.props.` lookup chain.

**Comment by bvaughn** [comment]
Hm, looks like there are some conflicts. Try rebasing?

File: source/Grid/Grid.js (hunk 1 above)
Comment (guilhermefloriani) [inline source/Grid/Grid.js]: It's done.

**Comment by bvaughn** [comment]
Excellent! Thank you! 😄

I'm going to hold off on merging this until #743 has been merged since that PR is pretty complicated and I don't want to cause any conflicts. Stay tuned!

File: source/Grid/Grid.js [hunk 2]
```diff
@@ -1041,17 +1041,23 @@ export default class Grid extends PureComponent {
...
     if (this._disablePointerEventsTimeoutId) {
-      clearTimeout(this._disablePointerEventsTimeoutId);
+      window.cancelAnimationFrame(this._disablePointerEventsTimeoutId)
     }
 
-    this._disablePointerEventsTimeoutId = setTimeout(
-      this._debounceScrollEndedCallback,
-      scrollingResetTimeInterval
-    );
+    const delay = () => {
```
Comment (TrySound) [inline source/Grid/Grid.js]: Allocating function on every scroll event can slow down performance.

File: source/Grid/Grid.js [hunk 3]
```diff
@@ -1037,20 +1038,32 @@ export default class Grid extends PureComponent {
   }
 
   /**
+   * Check if the difference between current time and the last scroll ended event is greater.
+   * than the scrollingResetTimeInterval prop, else schedule this function to execute again.
+   */
+  _delayScrollEnded() {
```
Comment (TrySound) [inline source/Grid/Grid.js]: Conflict can be solved If you will use class property here instead of bind
```
_delayScrollEnded = () => {
```

Comment (guilhermefloriani) [inline source/Grid/Grid.js]: Yes, I'm merging right now.
I was just getting a coffee :smile:

**Comment by guilhermefloriani** [comment]
I did the merge with these last changes on Grid component, but after the merge of this [commit](https://github.com/bvaughn/react-virtualized/commit/4f6f3d1be911551e4f15e076cc6d40835feef069), the build is broken:

```
...
source/Grid/Grid.jest.js -> dist/commonjs/Grid/Grid.jest.js
TypeError: source/Grid/Grid.js: Property value expected type of string but got null
    at Object.validate (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-types/lib/definitions/index.js:161:13)
    at validate (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-types/lib/index.js:505:9)
    at Object.builder (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-types/lib/index.js:466:7)
    at PluginPass.ObjectExpression (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-plugin-transform-es2015-duplicate-keys/lib/index.js:65:26)
    at newFn (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-traverse/lib/visitors.js:276:21)
    at NodePath._call (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-traverse/lib/path/context.js:76:18)
    at NodePath.call (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-traverse/lib/path/context.js:48:17)
    at NodePath.visit (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-traverse/lib/path/context.js:105:12)
    at TraversalContext.visitQueue (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-traverse/lib/context.js:150:16)
    at TraversalContext.visitSingle (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-traverse/lib/context.js:108:19)
    at TraversalContext.visit (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-traverse/lib/context.js:192:19)
    at Function.traverse.node (/home/guilherme/dev/js/test/react-virtualized/node_modules/babel-traverse/lib/index.js:114:17)
```

It is weird because no one warning or error was reported from flow or eslint.
I tried to checkout the `master` branch without my changes, but the problem already happens.

What do you think?

**Comment by TrySound** [comment]
I guess it's [https://github.com/brigand/babel-plugin-flow-react-proptypes](flow-react-proptypes)

**Comment by TrySound** [comment]
https://github.com/brigand/babel-plugin-flow-react-proptypes/pull/118

**Comment by bvaughn** [comment]
Although tests _pass_, this branch prints a bunch of warnings not present in master:
```
Warning: Can only update a mounted or mounting component. This usually means you called setState, replaceState, or forceUpdate on an unmounted component. This is a no-op.

Please check the code for the Grid component.
```

File: source/Grid/Grid.js [hunk 4]
```diff
@@ -1125,20 +1126,32 @@ export default class Grid extends PureComponent {
   }
 
   /**
+   * Check if the difference between current time and the last scroll ended event is greater.
+   * than the scrollingResetTimeInterval prop, else schedule this function to execute again.
+   */
+  _delayScrollEnded = () => {
+    const { scrollingResetTimeInterval } = this.props;
+    if (Date.now() - this._scrollDebounceStart >= scrollingResetTimeInterval) {
+      this._debounceScrollEndedCallback();
```
Comment (bvaughn) [inline source/Grid/Grid.js]: minor nit: I realize we weren't doing this before, and it's no big deal, but we might as well clear `this._disablePointerEventsTimeoutId` here.

File: source/Masonry/Masonry.js [hunk 5]
```diff
@@ -290,16 +291,28 @@ export default class Masonry extends PureComponent {
...
 
-  _debounceResetIsScrolling() {
+  /**
+   * Check if the difference between current time and the last scroll ended event is greater.
+   * than the scrollingResetTimeInterval prop, else schedule this function to execute again.
+   */
+  _delayScrollEnded() {
     const { scrollingResetTimeInterval } = this.props;
+    if (Date.now() - this._scrollDebounceStart >= scrollingResetTimeInterval) {
+      this._debounceScrollEndedCallback();
```
Comment (bvaughn) [inline source/Masonry/Masonry.js]: Ditto as for `Grid`

File: source/Grid/Grid.js (hunk 4 above)
Comment (guilhermefloriani) [inline source/Grid/Grid.js]: Yes, I agree that the pointer must be cleared, but this hasn't been doing on `_debounceScrollEndedCallback` function?

Comment (bvaughn) [inline source/Grid/Grid.js]: You're right 😄  Sorry, I was just skimming the diff and that callback wasn't visible. Sorry~ 👍

**Comment by jaredlunde** [comment]
@bvaughn @guilhermefloriani 

```
Warning: Can only update a mounted or mounting component. This usually means you called setState, replaceState, or forceUpdate on an unmounted component. This is a no-op.

Please check the code for the Grid component.
```

I'd guess that error is occurring because the following two lines should be `window.cancelAnimationFrame` rather than `window.clearTimeout`

- https://github.com/guilhermefloriani/react-virtualized/blob/c20bf15025cd511eef64702ac5b2ae89a117194e/source/Grid/Grid.js#L671
- https://github.com/guilhermefloriani/react-virtualized/blob/c20bf15025cd511eef64702ac5b2ae89a117194e/source/Masonry/Masonry.js#L117

**Comment by guilhermefloriani** [comment]
Nice @jaredlunde, now seems to be ok, the warnings not show up.
Sorry everyone, I didn't had time to work on this.

**Comment by jaredlunde** [comment]
One last place `requestAnimationFrame` needs to replace `setTimeout` is in the `WindowLoader` -> `onScroll` util. I have done so in a local branch and also replaced this kind of pattern which gets used multiple times with `requestAnimationTimeout` and `cancelAnimationTimeout` functions:

```js
    const delay = () => {
      if (Date.now() - this._scrollDebounceStart >= this.props.scrollingResetTimeInterval) {
        this._debounceScrollEndedCallback()
      } else {
        this._disablePointerEventsTimeoutId = window.requestAnimationFrame(delay)
      }
    }

    this._scrollDebounceStart = Date.now()
    this._disablePointerEventsTimeoutId = window.requestAnimationFrame(delay)
```

So this code snippet is reduced to:

```js
   this._disablePointerEventsTimeoutId = requestAnimationTimeout(
      this._debounceScrollEndedCallback, 
      this.props.scrollingResetTimeInterval
    )
```

And `window.cancelAnimationFrame(...)` becomes `cancelAnimationTimeout(...)`. This also removes the need for a `_scrollDebounceStart` property.

What's going to be the easiest way to review those changes? Does it make more sense to submit a separate PR altogether with this master branch or to submit a PR to @guilhermefloriani's branch?

**Comment by bvaughn** [comment]
Up to you 2 to work out. @guilhermefloriani could give you push access to his fork of RV and you could make your changes directly to it.

**Comment by guilhermefloriani** [comment]
I sent a invite to him.

File: source/WindowScroller/utils/onScroll.js [hunk 6]
```diff
@@ -1,3 +1,9 @@
+import requestAnimationTimeout, {
+  cancelAnimationTimeout
+} from '../../utils/requestAnimationTimeout'
```
Comment (TrySound) [inline source/WindowScroller/utils/onScroll.js]: Looks a bit weird. Maybe better to use only named exports? And then filename can be changed to `animationTimeout.js`

File: source/WindowScroller/utils/onScroll.js [hunk 7]
```diff
@@ -1,3 +1,9 @@
...
+
+
+
```
Comment (TrySound) [inline source/WindowScroller/utils/onScroll.js]: To much line breaks

**Comment by bvaughn** [comment]
This looks good! `Collection` still uses `setTimeout` but I don't think that's a very commonly used component so it's probably okay to update it later.

I think the only thing remaining would be to fallback to `setTimeout` for browsers that don't support `requestAnimationFrame` (eg IE 9). I've done this via f3beee0.

Thanks for your work on this issue!

**Comment by bvaughn** [comment]
This feature has been released in version [9.10.0](https://github.com/bvaughn/react-virtualized/blob/master/CHANGELOG.md#9100). Thank you for contributing!

**Comment by peisenmann** [comment]
Oh no, Collection is the one I'm using and experiencing this XD. Any odds there's a fix in the works?

**Pull request merged**